pkg/kudoctl/packages/verifier/template/verify_parameters.go

```diff
@@ -43,16 +49,98 @@ func (ParametersVerifier) Verify(pf *packages.Files) verifier.Result {
 
 func immutableParams(pf *packages.Files) verifier.Result {
 	res := verifier.NewResult()
-	for _, value := range pf.Params.Parameters {
-		if value.IsImmutable() {
-			if !value.HasDefault() && !value.IsRequired() {
-				res.AddParamError(value.Name, "is immutable but is not marked as required or has a default value")
+	for _, p := range pf.Params.Parameters {
+		if p.IsImmutable() {
+			if !p.HasDefault() && !p.IsRequired() {
+				res.AddParamError(p.Name, "is immutable but is not marked as required or has a default value")
+			}
+		}
+	}
+	return res
+}
+
+func paramDefaults(pf *packages.Files) verifier.Result {
+	res := verifier.NewResult()
+	for _, p := range pf.Params.Parameters {
+		if p.HasDefault() {
+			defaultAsString, ok := p.Default.(string)
+			if !ok {
+				res.AddParamError(p.Name, fmt.Sprintf("failed to convert default value %q to string", p.Default))
+				continue
+			}
+			if err := kudoapi.ValidateParameterValueForType(p.Type, defaultAsString); err != nil {
+				res.AddParamError(p.Name, fmt.Sprintf("has an invalid default value: %v", err))
+			}
+			if p.IsEnum() {
+				if err := kudoapi.ValidateParameterValueForEnum(p.EnumValues(), defaultAsString); err != nil {
+					res.AddParamError(p.Name, fmt.Sprintf("has an invalid default value: %v", err))
+				}
+			}
+		}
+	}
+	return res
+}
+
+func enumParams(pf *packages.Files) verifier.Result {
+	res := verifier.NewResult()
+	for _, p := range pf.Params.Parameters {
+
+		if p.IsEnum() {
+			if len(p.EnumValues()) == 0 {
+				res.AddParamError(p.Name, "is an enum but has no allowed values")
+				continue
+			}
+			for _, enumVal := range p.EnumValues() {
+
+				if err := kudoapi.ValidateParameterValueForType(p.Type, enumVal); err != nil {
+					res.AddParamError(p.Name, fmt.Sprintf("has an invalid enum value: %v", err))
+				}
 			}
 		}
 	}
 	return res
 }
 
+func metadata(pf *packages.Files) verifier.Result {
+	res := verifier.NewResult()
+	for _, p := range pf.Params.Parameters {
+		if p.Group != "" {
+			if strings.Contains(p.Group, "/") {
+				res.AddParamError(p.Name, "has a group with invalid character '/'")
+			}
+		}
+		if p.IsAdvanced() && !p.HasDefault() && p.IsRequired() {
+			res.AddParamError(p.Name, "is marked as advanced, but also as required and has no default. An advanced parameter must either be optional or have a default value")
+		}
+	}
+	return res
+}
+
+func paramGroups(pf *packages.Files) verifier.Result {
+	res := verifier.NewResult()
+
+	groups := map[string]packages.Group{}
+	for _, g := range pf.Params.Groups {
+		if strings.Contains(g.Name, "/") {
+			res.AddGroupError(g.Name, "contains invalid character '/'")
+		}
+		if _, ok := groups[g.Name]; ok {
+			res.AddGroupError(g.Name, "is duplicated")
+		}
+		groups[g.Name] = g
+	}
+
+	for _, p := range pf.Params.Parameters {
+		if p.Group != "" {
+			if _, ok := groups[p.Group]; !ok {
+				res.AddParamWarning(p.Name, "has a group that is not defined in the group section")
```

Comment:
why is this just warn and not err?

Reply:
Because I think it's ok to not define the groups - in this case they'd have to be displayed without a description and displayName, simply with the group name.

Reply:
hmm I thought that a potential UI will read all the groups and then render params in those groups. But then most of the params are not expected to be in any group or be in group but that group not defined? That sounds like a lot of edge cases a potential UI have to handle 🤔 

Just a thought...

Reply:
Yeah... Valid point. I guess it'd be acceptable if people want to use a group that they have to define the group first. I'll make it mandatory.